core/src/main/java/hudson/model/DirectoryBrowserSupport.java

```diff
@@ -339,17 +355,31 @@ private void serveFile(StaplerRequest req, StaplerResponse rsp, VirtualFile root
             // pseudo file name to let the Stapler set text/plain
             rsp.serveFile(req, in, lastModified, -1, length, "plain.txt");
         } else {
-            String csp = SystemProperties.getString(DirectoryBrowserSupport.class.getName() + ".CSP", DEFAULT_CSP_VALUE);
-            if (!csp.trim().equals("")) {
-                // allow users to prevent sending this header by setting empty system property
-                for (String header : new String[]{"Content-Security-Policy", "X-WebKit-CSP", "X-Content-Security-Policy"}) {
-                    rsp.setHeader(header, csp);
+            if (shouldRedirectToResourceDomain(req)) {
+                // redirect to second domain
+                rsp.sendRedirect(302, ExtensionList.lookupSingleton(ResourceDomainRootAction.class).getRedirectUrl(resourceToken, req.getRestOfPath()));
+            } else {
+                if (!ResourceDomainConfiguration.isResourceRequest(req)) {
+                    // if we're serving this from the main domain, set CSP headers
+                    String csp = SystemProperties.getString(DirectoryBrowserSupport.class.getName() + ".CSP", DEFAULT_CSP_VALUE);
+                    if (!csp.trim().equals("")) {
+                        // allow users to prevent sending this header by setting empty system property
+                        for (String header : new String[]{"Content-Security-Policy", "X-WebKit-CSP", "X-Content-Security-Policy"}) {
+                            rsp.setHeader(header, csp);
+                        }
+                    }
                 }
+                rsp.serveFile(req, in, lastModified, -1, length, baseFile.getName());
             }
-            rsp.serveFile(req, in, lastModified, -1, length, baseFile.getName() );
         }
     }
-    
+
+    private boolean shouldRedirectToResourceDomain(StaplerRequest req) {
+        boolean isResourceDomainRequest = ResourceDomainConfiguration.isResourceRequest(req);
+        boolean hasConfiguredResourceDomain = ResourceDomainConfiguration.isResourceDomainConfigured();
+        return resourceToken != null && !isResourceDomainRequest && hasConfiguredResourceDomain;
```

Comment:
Better to inline the vars so we can short-circuit some method calls.

Reply:
They're unnecessary anyway, so will just remove them.